subsys/bluetooth/mesh/prov.c

```diff
@@ -122,57 +124,166 @@ static int check_input_auth(bt_mesh_input_action_t input, uint8_t size)
-		/* pull base-36 digits: */
-		int idx = value % 36;
+	bt_rand(str, size);
 
-		value = value / 36;
-		str[i] = characters[idx];
+	for (int i = 0; i < size; i++) {
+		str[i] = characters[(uint8_t)str[i] % (uint8_t)36];
 	}
 
-	str[size] = '\0';
-
 	memcpy(bt_mesh_prov_link.auth, str, size);
 	memset(bt_mesh_prov_link.auth + size, 0,
 			sizeof(bt_mesh_prov_link.auth) - size);
 }
 
-static uint32_t get_auth_number(bt_mesh_output_action_t output,
-		bt_mesh_input_action_t input, uint8_t size)
+/* Computes 10^n - 1 in little endian array */
+static void compute_pow10_minus1(uint8_t n, uint8_t *out, size_t *out_len)
+{
+	out[0] = 1; /* Start with value 1 at LSB */
+	size_t len = 1;
+
+	/* Compute 10^n */
+	for (uint8_t i = 0; i < n; ++i) {
+		uint16_t carry = 0;
+
+		for (size_t j = 0; j < len; ++j) {
+			uint16_t prod = out[j] * 10 + carry;
+
+			out[j] = prod & 0xFF;
+			carry = prod >> 8;
+		}
+
+		if (carry > 0 && len < MAX_NUMERIC_OOB_BYTES) {
+			out[len] = carry;
+			len++;
+		}
+	}
+
+	/* Subtract 1 from the result (10^n - 1) */
+	size_t j = 0;
+
+	while (j < len) {
+		if (out[j] > 0) {
+			out[j] -= 1;
+			break;
+		}
+		out[j] = 0xFF;
+		j++;
+	}
+
+	/* Adjust length if highest byte becomes 0 */
+	while (len > 1 && out[len - 1] == 0) {
+		len--;
+	}
+
+	*out_len = len;
+}
+
+static void generate_random_byte_shift(uint8_t max_inclusive, uint8_t *out)
+{
+	uint8_t byte;
+
+	if (max_inclusive == 0) {
+		*out = 0;
+		return;
+	}
+
+	bt_rand(&byte, 1);
+	while (byte > max_inclusive) {
+		byte >>= 1;
+	}
+
+	*out = byte;
+}
+
+static void generate_random_below_pow10(uint8_t n, uint8_t *output, size_t *out_len)
+{
+	uint8_t max_val[MAX_NUMERIC_OOB_BYTES] = {0};
+	size_t max_len = 0;
+	int free_fill = 0;
+
+	compute_pow10_minus1(n, max_val, &max_len);
+
+	/* Generate random number less than max_val, starting from MSB in little-endian */
+	for (int i = max_len - 1; i >= 0; --i) {
+		generate_random_byte_shift(max_val[i], &output[i]);
+		if (output[i] < max_val[i]) {
+			free_fill = i;
+			break;
+		}
+	}
+
+	/* Fill remaining lower significant bytes with random data */
+	if (free_fill > 0) {
+		bt_rand(&output[0], free_fill);
+	}
+
+	/* Ensure the result is not all zero */
+	int all_zero = 1;
+
+	for (size_t i = 0; i < max_len; ++i) {
+		if (output[i] != 0) {
+			all_zero = 0;
+			break;
+		}
+	}
+
+	if (all_zero) {
+		output[0] = 1;
+	}
+
+	*out_len = max_len;
+}
+
+static void get_auth_number(uint8_t *rand_bytes, size_t *size)
 {
-	const uint32_t divider[PROV_IO_OOB_SIZE_MAX] = { 10, 100, 1000, 10000,
-			100000, 1000000, 10000000, 100000000 };
 	uint8_t auth_size = bt_mesh_prov_auth_size_get();
-	uint32_t num = 0;
 
-	bt_rand(&num, sizeof(num));
+	generate_random_below_pow10(*size, rand_bytes, size);
+	sys_memcpy_swap(bt_mesh_prov_link.auth + (auth_size - *size), rand_bytes, *size);
+	memset(bt_mesh_prov_link.auth, 0, auth_size - *size);
+}
+
+void bt_mesh_binary_to_ascii_digits(const uint8_t *in, size_t in_len, char *out_str)
```

Comment:
I'm not sure why this function needs to be part of stack code. As far as I see it is only used for printing oob data in a readable text and only used in shell and in test. I suggest to move it to shell and test code.

Also, since it is only used for printing, I think it can be simplified using `%c`. E.g. this is how logging subsystem prints hexdump: https://github.com/zephyrproject-rtos/zephyr/blob/0df9c5fb1f21df1c34daf12664d38e77e729af6d/subsys/logging/log_minimal.c#L42-L50

Reply:
My initial implementation was exactly what you requested. However, Sonar checker banned it due to huge percentage of similarity. That's the truth, since the functions are identical in shell and in test. 
The checker is quite stupid in understanding purposes. Nevertheless, it banns PR. Instead, I added the single one here. I do not like it either. However, it is not a big problem since the function should be cut out by linker if this is not used.

Any better ideas are welcome.

Reply:
We should try to suppress Sonar I think. I'm still confident that if this function doesn't belong to stack, it should not be there.